devito/arguments.py

```diff
@@ -0,0 +1,183 @@
+from cached_property import cached_property
+import cgen as c
+import numpy as np
+
+
+class RuntimeArgProvider(object):
+
+    @property
+    def rtargs(self):
+        raise NotImplemented()
+
+
+class RuntimeArgument(object):
+
+    def __init__(self, name, source, default_value=None):
+        self.name = name
+        self.source = source
+        self._value = self._default_value = default_value
+
+    @property
+    def value(self):
+        return self._value
+
+    @property
+    def ready(self):
+        return self._value is not None
+
+    @property
+    def decl(self):
+        raise NotImplemented()
+
+    def reset(self):
+        self._value = self._default_value
+
+    def verify(self, kwargs):
+        raise NotImplemented()
+
+
+class ScalarArgument(RuntimeArgument):
+
+    def __init__(self, name, source, reducer, default_value=None):
+        super(ScalarArgument, self).__init__(name, source, default_value)
+        self.reducer = reducer
+
+    @property
+    def decl(self):
+        return c.Value('const int', self.name)
+
+    @property
+    def dtype(self):
+        return self.source.dtype
+
+    def verify(self, value):
+        # Assuming self._value was initialised as appropriate for the reducer
+        if value is not None:
+            if self._value is not None:
+                self._value = self.reducer(self._value, value)
+            else:
+                self._value = value
+        return self._value is not None
+
+
+class TensorArgument(RuntimeArgument):
+
+    def __init__(self, name, source, dtype):
+        super(TensorArgument, self).__init__(name, source)
+        self.dtype = dtype
+        self._value = self._default_value = self.source
+
+    @property
+    def value(self):
+        if isinstance(self._value, np.ndarray):
+            return self._value
+        else:
+            return self._value.data
```

Comment:
Let's do it in the next one and make sure for now that this data is `None` for OperatorForeign.